File: Lecture15-Filtering/main.cpp
#include <opencv4/opencv2/opencv.hpp>

using namespace cv;

void filter2d()
{
    Mat img = imread("/home/berkan/CLionProjects/OpenCV/Ders15_Filtreleme/castle.jpg", IMREAD_GRAYSCALE);
    threshold(img, img, 128, 255, THRESH_BINARY);
    imshow("Original", img);

    float data[] = {1, 1, 1,
                    1, -7, 1,
                    1, 1, 1};
    Mat kernel(3, 3, CV_32F, data);
    filter2D(img, img, -1, kernel);
    imshow("Castle", img);
    waitKey(0);
    destroyAllWindows();
}

void sepFilter()
{
    Mat img = imread("/home/berkan/CLionProjects/OpenCV/Ders15_Filtreleme/castle.jpg", IMREAD_GRAYSCALE);
    threshold(img, img, 128, 255, THRESH_BINARY);
    imshow("Original", img);

    float dataX[] = {-1, -2, 6, -2, -1};
    float dataY[] = {0, 0, 1, 0, 0};
    Mat kernelX(1, 5, CV_32F, dataX);
    Mat kernelY(5, 1, CV_32F, dataY);
    sepFilter2D(img, img, -1, kernelX, kernelY);
    imshow("Sep Filter", img);
    waitKey(0);
}

void findEdge()
{
    Mat img = imread("/home/berkan/CLionProjects/OpenCV/Ders15_Filtreleme/castle.jpg", IMREAD_GRAYSCALE);
    threshold(img, img, 128, 255, THRESH_BINARY);
    imshow("Original", img);

    float data[] = {1, 1, 1,
                    1, 8, 1,
                    1, 1, 1};
    Mat kernel(3, 3, CV_32F, data);
    filter2D(img, img, -1, kernel);
    imshow("Finding Edge", img);
    waitKey(0);
}

int main() {
    //filter2d();
    //sepFilter();
    findEdge();
    return 0;
}

File: Lecture15-Filtering/main.py
import cv2
import numpy as np

img = cv2.imread("Kale.jpg", cv2.IMREAD_GRAYSCALE)
cv2.threshold(img, 128, 255, cv2.THRESH_BINARY)
cv2.imshow("Orijinal", img)

def filter2D():
    kernel = np.array([[1,1,1],
                      [1,-7,1],
                      [1,1,1]], np.float32)
    filterImg = cv2.filter2D(img, cv2.CV_32F, kernel=kernel)
    cv2.imshow("2D Filter", filterImg)
    cv2.waitKey(0)
    cv2.destroyAllWindows()


def kenarBulma():
    kernel = np.array([[1,1,1],
                      [1,8,1],
                      [1,1,1]], np.float32)
    filterImg = cv2.filter2D(img, cv2.CV_32F, kernel)
    cv2.imshow("Kenar Bulma", filterImg)
    cv2.waitKey(0)
    cv2.destroyAllWindows()

def sepFilter():
     dataX = [-1, -2, 6, -2, -1]
     dataY = [0, 0, 1, 0, 0]

     kernelX = np.array((1, 5), cv2.CV_32F, dataX)
     kernelY = np.array((5, 1), cv2.CV_32F, dataY)

     cv2.sepFilter2D(img, -1, kernelX, kernelY)

     cv2.imshow("Sep Filter", img)

     cv2.waitKey(0)
     cv2.destroyAllWindows()

if __name__ == '__main__':
    kenarBulma()
    cv2.waitKey(0)
    cv2.destroyAllWindows()